pkg/controller/direct/directbase/directbase_controller.go

```diff
@@ -0,0 +1,393 @@
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	logger.Info("successfully finished reconcile", "resource", request.NamespacedName, "time to next reconciliation", jitteredPeriod)
+	return reconcile.Result{RequeueAfter: jitteredPeriod}, nil
+}
+
+func (r *reconcileContext) doReconcile(ctx context.Context, u *unstructured.Unstructured) (requeue bool, err error) {
+	logger := log.FromContext(ctx)
+
+	adapter, err := r.Reconciler.model.AdapterForObject(ctx, u)
+	if err != nil {
+		return false, r.handleUpdateFailed(ctx, u, err)
+	}
+
+	defer execution.RecoverWithInternalError(&err)
+	if !u.GetDeletionTimestamp().IsZero() {
+		if !k8s.HasFinalizer(u, k8s.ControllerFinalizerName) {
+			// Resource has no controller finalizer; no finalization necessary
+			return false, nil
+		}
+		if k8s.HasFinalizer(u, k8s.DeletionDefenderFinalizerName) {
+			// deletion defender has not yet been finalized; requeuing
+			logger.Info("deletion defender has not yet been finalized; requeuing", "resource", k8s.GetNamespacedName(u))
+			return true, nil
+		}
+		if !k8s.HasAbandonAnnotation(u) {
+			if _, err := adapter.Delete(ctx); err != nil {
+				if !errors.Is(err, kcciamclient.NotFoundError) && !k8s.IsReferenceNotFoundError(err) {
+					if unwrappedErr, ok := lifecyclehandler.CausedByUnresolvableDeps(err); ok {
+						logger.Info(unwrappedErr.Error(), "resource", k8s.GetNamespacedName(u))
+						resource, err := toK8sResource(u)
+						if err != nil {
+							return false, fmt.Errorf("error converting k8s resource while handling unresolvable dependencies event: %w", err)
+						}
+						// Requeue resource for reconciliation with exponential backoff applied
+						return true, r.Reconciler.HandleUnresolvableDeps(ctx, resource, unwrappedErr)
+					}
+					return false, r.handleDeleteFailed(ctx, u, err)
+				}
+			}
+		}
+		return false, r.handleDeleted(ctx, u)
+	}
+
+	existsAlready, err := adapter.Find(ctx)
+	if err != nil {
+		if unwrappedErr, ok := lifecyclehandler.CausedByUnresolvableDeps(err); ok {
+			logger.Info(unwrappedErr.Error(), "resource", k8s.GetNamespacedName(u))
+			return r.handleUnresolvableDeps(ctx, u, unwrappedErr)
+		}
+		return false, r.handleUpdateFailed(ctx, u, err)
+	}
+	k8s.EnsureFinalizers(u, k8s.ControllerFinalizerName, k8s.DeletionDefenderFinalizerName)
+
+	// set the etag to an empty string, since IAMPolicy is the authoritative intent, KCC wants to overwrite the underlying policy regardless
+	//policy.Spec.Etag = ""
+
+	if !existsAlready {
+		if err := adapter.Create(ctx, u); err != nil {
+			if unwrappedErr, ok := lifecyclehandler.CausedByUnresolvableDeps(err); ok {
+				logger.Info(unwrappedErr.Error(), "resource", k8s.GetNamespacedName(u))
+				return r.handleUnresolvableDeps(ctx, u, unwrappedErr)
+			}
+			return false, r.handleUpdateFailed(ctx, u, fmt.Errorf("error creating: %w", err))
+		}
+	} else {
+		if _, err = adapter.Update(ctx); err != nil {
```

Comment:
Looks like we don't do a diff for the entire spec but only check the fields Update() function cares about?
Not sure if it is applicable to this resource, but overall the immutability checks will be necessary for invalid changes.

Reply:
Added TODO